### ISSUE TITLE
[BUILD] Compile WITH_OTLP failed

### ISSUE THREAD

**Issue by ivgotcrazy** [open-telemetry/opentelemetry-cpp#2095]
**Describe your environment** 
Linux CentOS7 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
gcc 12.2.0

**Steps to reproduce**
1) git clone master
2) cmake3 -DBUILD_TESTING=OFF -DCMAKE_CXX_STANDARD=17 -DWITH_OTLP=ON ..

-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building for architecture ARCH=x64
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
CMake Warning at /usr/share/cmake3/Modules/FindProtobuf.cmake:500 (message):
  Protobuf compiler version 22.2 doesn't match library version 4.22.2
Call Stack (most recent call first):
  CMakeLists.txt:333 (find_package)

-- Found Protobuf: /usr/local/lib64/libprotobuf.a;-lpthread (found version "4.22.2") 
-- PROTOBUF_PROTOC_EXECUTABLE=/usr/local/bin/protoc
-- opentelemetry-proto dependency satisfied by: git submodule
-- Performing Test check_cxx_compiler_flag_-Wno-type-limits
-- Performing Test check_cxx_compiler_flag_-Wno-type-limits - Success
-- Performing Test check_cxx_compiler_flag_-Wno-deprecated-declarations
-- Performing Test check_cxx_compiler_flag_-Wno-deprecated-declarations - Success
-- Performing Test check_cxx_compiler_flag_-Wno-unused-parameter
-- Performing Test check_cxx_compiler_flag_-Wno-unused-parameter - Success
-- gRPC_CPP_PLUGIN_EXECUTABLE=/usr/local/bin/grpc_cpp_plugin
Building with nostd types...
PATH:/root/opentelemetry-cpp/examples/grpc/protos:/root/opentelemetry-cpp/examples/grpc/protos/messages.proto
-- Configuring done
-- Generating done
-- Build files have been written to: /root/opentelemetry-cpp/build

3) make

**What is the expected behavior?**
success

**What is the actual behavior?**
failed

**Additional context**
[error.txt](https://github.com/open-telemetry/opentelemetry-cpp/files/11220281/error.txt)

**Comment by marcalff** [comment]
I suspect that `-DCMAKE_CXX_STANDARD=17` in fact requires `-DWITH_STL` in practice, which should be documented and/or enforced in the makefile.

Here, the compiler which supports C++17 compiles with replacement headers in the nostd namespace, which are meant to provide C++11 features to pre-C++11 compilers ... causing the errors reported.

**Comment by ivgotcrazy** [comment]
thanks a lot, it seems to work.

**Comment by lidavidm** [comment]
We ran into this on the [conda-forge package](https://github.com/conda-forge/cpp-opentelemetry-sdk-feedstock/pull/48) when Protobuf got bumped. `WITH_STL` mostly fixes things, but it seems Protobuf 4.23 also removed `google/protobuf/stubs/stringpiece.h` and `google/protobuf/stubs/strutil.h` which causes issues for otlp_http_client.cc. (And removing the includes isn't enough, since that file calls `google::protobuf::Base64Escape`; that appears to no longer be in the Protobuf API.)

**Comment by h-vetinari** [comment]
> but it seems Protobuf 4.23 also removed `google/protobuf/stubs/stringpiece.h` and `google/protobuf/stubs/strutil.h` which causes issues for otlp_http_client.cc

The commit for this is https://github.com/protocolbuffers/protobuf/commit/a3c8e2deb05186334b7ee8c1174f44802e38b43d, and it seems to be due to the switch to abseil, e.g.:
```diff
   static const char kSuffix[] = "_FIELD_NUMBER";
   if (PyString_AsStringAndSize(name, &attr, &attr_size) >= 0 &&
-      HasSuffixString(StringPiece(attr, attr_size), kSuffix)) {
+      HasSuffixString(absl::string_view(attr, attr_size), kSuffix)) {
     std::string field_name(attr, attr_size - sizeof(kSuffix) + 1);
     LowerString(&field_name);
```
